Spigot-API-Patches/0132-Custom-plugin-advancement-API.patch

```diff
@@ -0,0 +1,260 @@
+new file mode 100644
+index 00000000..9d969fe7
+--- /dev/null
++++ b/src/main/java/org/bukkit/advancement/AdvancementData.java
+@@ -0,0 +1,157 @@
++package org.bukkit.advancement;
++
++import org.bukkit.NamespacedKey;
++import org.bukkit.inventory.ItemStack;
++
++import javax.annotation.Nullable;
++
++public class AdvancementData {
++    private final NamespacedKey parent;
++    private final AdvancementDisplay display;
++    private final AdvancementRewards rewards;
++    private final NamespacedKey key;
++
++    public AdvancementData(@Nullable NamespacedKey parent, AdvancementDisplay display, AdvancementRewards rewards, NamespacedKey key) {
++        this.parent = parent;
++        this.display = display;
++        this.rewards = rewards;
++        this.key = key;
++    }
++
++    public NamespacedKey getKey() {
++        return key;
++    }
++
++    /**
++     * The optional parent advancement of this advancement (does not apply for the root advancement).
++     * @return The parent advancement if present otherwise null
++     */
++    @Nullable
++    public NamespacedKey getParent() {
++        return parent;
++    }
++
++    /**
++     * The optional display data. Controls how the advancement looks for the client.
++     * @return Display data if present otherwise null
++     */
++    @Nullable
++    public AdvancementDisplay getDisplay() {
++        return display;
++    }
++
++    /**
++     * Required rewards object which gives rewards when a player completes the advancement.
++     * Can be used to define a custom function for executing commands (which can then execute code..)
++     * @return The rewards
++     */
++    public AdvancementRewards getRewards() {
++        return rewards;
++    }
++
++    public static class AdvancementRewards {
++        private int experience;
++        private NamespacedKey[] loot;
++        private NamespacedKey[] recipes;
++        @Nullable
++        private NamespacedKey customFunction;
++
++        public AdvancementRewards(int experience) {
++            this(experience, new NamespacedKey[0], new NamespacedKey[0], null);
++        }
++
++        public AdvancementRewards(int experience, NamespacedKey customFunction) {
++            this(experience, new NamespacedKey[0], new NamespacedKey[0], customFunction);
++        }
++
++        public AdvancementRewards(NamespacedKey customFunction) {
++            this(0, new NamespacedKey[0], new NamespacedKey[0], customFunction);
++        }
++
++        public AdvancementRewards(int experience, NamespacedKey[] loot, NamespacedKey[] recipes, @Nullable NamespacedKey customFunction) {
++            this.experience = experience;
++            this.loot = loot;
++            this.recipes = recipes;
++            this.customFunction = customFunction;
++        }
++
++        public int getExperience() {
++            return experience;
++        }
++
++        public NamespacedKey[] getLoot() {
```

Comment:
Bukkit API uses collections, not arrays

Reply:
Should be a Set, but also maybe worth returning LootTable directly.

Reply:
it internally uses an array of `MinecraftKey`s

Reply:
The common usage of this API would more suit a Set, as aikar stated; No need to stick to vanilla implementation detail if we can expose something that is more suitable for API

Reply:
oh yes,  I was referring to loot table though

Reply:
Well im ok with returning just the keys, and let plugins access the loot table themselves.

i dont know how heavy getting a loot table is, so i guess it is good to let them choose if they need it or not.

Set is best because it doesnt make sense to ever return the same key twice in the collection.

Reply:
:ok_hand: